src/back/glsl/mod.rs

```diff
@@ -514,8 +542,9 @@ impl<'a, W: Write> Writer<'a, W> {
     /// - If type is either a image or sampler
     /// - If it's an Array with a [`ArraySize::Constant`](crate::ArraySize::Constant) with a
     /// constant that isn't [`Uint`](crate::ConstantInner::Uint)
-    fn write_type(&mut self, ty: Handle<Type>) -> BackendResult {
-        match self.module.types[ty].inner {
+    /// - If the type is a struct and the `handle` is None
+    fn write_type(&mut self, ty: &TypeInner, handle: Option<Handle<Type>>) -> BackendResult {
```

Comment:
Technically this should accept either a handle, or a reference to the type inner, so it could work with something like the thing that `get_handle()` returns: `Result<Handle<crate::Type>, &crate::TypeInner>`
That would allow call sites to skip doing the silly thing as `(&self.module.types[base].inner, Some(base))` (that would simply be `Ok(base)`. For clearest semantics we could define our own enum instead of using the `Result`:
```rust
enum ExpressionType<'a> {
  Handle(Handle<Expression>),
  Value(&'a TypeInner),
}
```